star-rx72n-firmware/lib/rx_fec/src/rx_fec.c

```diff
@@ -0,0 +1,461 @@
+/**
+ * @file rx_fec.c
+ * @brief Forward Error Correction (FEC) Codec Implementation
+ *
+ * Implements NASA-standard K=7 convolutional encoder and Viterbi decoder.
+ * This is a C port of star-gateway/internal/fec/ for bit-exact compatibility.
+ *
+ * STAR Project - Texas A&M University
+ * December 2025
+ */
+
+#include "rx_fec.h"
+
+#include <string.h>
+
+/* =============================================================================
+ * Private Constants
+ * =============================================================================
+ */
+
+/** Maximum path metric value for initialization */
+#define MAX_PATH_METRIC 0x7FFFFFFF
+
+/* =============================================================================
+ * Private Helper Functions
+ * =============================================================================
+ */
+
+/**
+ * @brief Calculate parity (XOR of all bits) of a byte
+ *
+ * @param[in] x Input byte
+ * @return 0 if even parity, 1 if odd parity
+ */
+static uint8_t internal_parity(uint8_t x)
+{
+    x ^= x >> 4;
+    x ^= x >> 2;
+    x ^= x >> 1;
+    return x & 1;
+}
+
+/**
+ * @brief Set a bit at specified index in output buffer (MSB first)
+ *
+ * @param[out] output Output buffer
+ * @param[in]  bit_idx Bit index (0 = MSB of first byte)
+ * @param[in]  value Bit value (0 or 1)
+ */
+static void internal_set_output_bit(uint8_t *output, size_t bit_idx,
+                                    uint8_t value)
+{
+    size_t byte_idx = bit_idx / 8;
+    size_t bit_pos  = 7 - (bit_idx % 8); /* MSB first */
+    if (value != 0) {
+        output[byte_idx] |= (uint8_t)(1 << bit_pos);
+    }
+}
+
+/**
+ * @brief Get a bit at specified index from buffer (MSB first)
+ *
+ * @param[in] data Input buffer
+ * @param[in] bit_idx Bit index (0 = MSB of first byte)
+ * @return Bit value (0 or 1)
+ */
+static uint8_t internal_get_bit(const uint8_t *data, size_t bit_idx)
+{
+    size_t byte_idx = bit_idx / 8;
+    size_t bit_pos  = 7 - (bit_idx % 8); /* MSB first */
+    return (data[byte_idx] >> bit_pos) & 1;
+}
+
+/**
+ * @brief Encode a single bit and update encoder state
+ *
+ * @param[in,out] state Current encoder state (6-bit shift register)
+ * @param[in]     input_bit Input bit (0 or 1)
+ * @param[out]    out0 First output bit (G1)
+ * @param[out]    out1 Second output bit (G2)
+ */
+static void internal_encode_bit(uint8_t *state, uint8_t input_bit,
+                                uint8_t *out0, uint8_t *out1)
+{
+    /* Shift in the new bit (input is MSB of the combined state) */
```

Comment:
Since we are modifing the pointer a comment should be left in the doxygen to express that.